tracing/src/main/java/com/palantir/tracing/DeferredTracer.java

```diff
@@ -136,61 +134,30 @@ CloseableTrace withTrace() {
             Tracer.fastStartSpan(operation);
         }
 
-        if (isObservable && metadata != null && !metadata.isEmpty()) {
-            return new TaggedCloseableTrace(originalTrace, metadata);
-        } else {
-            return originalTrace.map(CLOSEABLE_TRACE_FUNCTION).orElse(DefaultCloseableTrace.INSTANCE);
-        }
-    }
-
-    private enum NopCloseableTrace implements CloseableTrace {
-        INSTANCE;
-
-        @Override
-        public void close() {}
-    }
-
-    private enum DefaultCloseableTrace implements CloseableTrace {
-        INSTANCE;
-
-        @Override
-        public void close() {
-            Tracer.fastCompleteSpan();
-            if (Tracer.hasTraceId()) {
-                Tracer.getAndClearTrace();
+        return new CloseableTracer() {
+            @Override
+            public void close() {
+                if (metadata != null) {
+                    Tracer.fastCompleteSpan(metadata);
+                } else {
+                    Tracer.fastCompleteSpan();
+                }
+                if (originalTrace.isPresent()) {
+                    Tracer.setTrace(originalTrace.get());
+                } else if (Tracer.hasTraceId()) {
+                    Tracer.getAndClearTrace();
+                }
```

Comment:
not exactly a new problem, but now that we're exposing `CloseableTracer` to consumers outside the library's control, do we need to make `close()` idempotent?

Reply:
What do you mean? We already expose `CloseableTracer` to consumers. Today consumers can write:

```java
try (CloseableTracer span = CloseableTracer.startSpan("operation")) {
    ...
}
```

Reply:
I agree we should make it idempotent, but I think that's outside the scope of this PR.